crates/grafbase-sdk/src/host_io/postgres/query.rs

```diff
@@ -0,0 +1,232 @@
+    Connection(&'a Connection),
+    Transaction(&'a Transaction),
+}
+
+impl<'a> From<&'a Connection> for ConnectionLike<'a> {
+    fn from(connection: &'a Connection) -> Self {
+        ConnectionLike::Connection(connection)
+    }
+}
+
+impl<'a> From<&'a Transaction> for ConnectionLike<'a> {
+    fn from(transaction: &'a Transaction) -> Self {
+        ConnectionLike::Transaction(transaction)
+    }
+}
+
+impl ConnectionLike<'_> {
+    pub fn query<'a>(
+        &'a self,
+        query: &'a str,
+        params: (&[wit::PgBoundValue], &[wit::PgValue]),
+    ) -> Result<Vec<wit::PgRow>, String> {
+        match self {
+            ConnectionLike::Connection(connection) => connection.0.query(query, params),
+            ConnectionLike::Transaction(transaction) => transaction.inner.query(query, params),
+        }
+    }
+
+    pub fn execute<'a>(
+        &'a self,
+        query: &'a str,
+        params: (&[wit::PgBoundValue], &[wit::PgValue]),
+    ) -> Result<u64, String> {
+        match self {
+            ConnectionLike::Connection(connection) => connection.0.execute(query, params),
+            ConnectionLike::Transaction(transaction) => transaction.inner.execute(query, params),
+        }
+    }
+}
+
+/// A query builder for constructing and executing SQL queries.
+///
+/// This struct provides a fluent interface for building SQL queries with
+/// parameter binding, and methods to execute those queries or fetch their results.
+#[derive(Clone, Debug)]
+pub struct Query<'a> {
+    pub(crate) connection: ConnectionLike<'a>,
+    pub(crate) query: &'a str,
+    pub(crate) values: Vec<wit::PgBoundValue>,
+    pub(crate) value_tree: wit::PgValueTree,
+}
+
+impl Query<'_> {
+    /// Binds a value to the query as a parameter.
+    ///
+    /// This method adds a value to be used as a parameter in the SQL query.
+    /// The value will be properly escaped and converted to the appropriate PostgreSQL type.
+    ///
+    /// # Parameters
+    /// * `value` - Any value that implements the `DatabaseType` trait
+    ///
+    /// # Returns
+    /// The query builder for method chaining
+    pub fn bind(mut self, value: impl DatabaseType) -> Self {
+        let DatabaseValue { value, array_values } = value.into_bound_value();
+
+        self.values.push(value);
+
+        if let Some(array_values) = array_values {
+            self.value_tree.extend(array_values);
+        }
+
+        self
+    }
+
+    /// Executes the SQL query with the bound parameters.
+    ///
+    /// This method sends the query to the database server and returns the number of rows affected.
+    /// For INSERT, UPDATE, or DELETE statements, this represents the number of rows modified.
+    /// For other statements, the meaning of the return value depends on the specific operation.
+    ///
+    /// # Returns
+    /// The number of rows affected by the query, or an error message if the execution failed
+    pub fn execute(self) -> Result<u64, String> {
```

Comment:
Would consider using `Result<..., SdkError>` everywhere to hide the fact it's a `String` and leave your some margin the future for changes.